language/move-binary-format/src/normalized.rs

```diff
@@ -272,3 +304,64 @@ impl Function {
         (name, f)
     }
 }
+
+impl From<&TypeTag> for Type {
```

Comment:
Looks like there is a lot of cloning cloning on. Why not just make this `From<TypeTag>` and force the cloning on the outside?

Reply:
It's mostly the `Box<T>` that makes things worse...you can't do an unbox so you kinda have to either deal with reference or clone the box out.

Reply:
I don't really follow. It should be possible to write it for an owned value, box or no box

Reply:
I think I'm not being clear so, this should work? (haven't tested it)
```rust
impl From<TypeTag> for Type {    
    fn from(ty: TypeTag) -> Type {
        use Type::*;
        match ty {
            TypeTag::Bool => Bool,
            TypeTag::U8 => U8,
            TypeTag::U64 => U64,
            TypeTag::U128 => U128,
            TypeTag::Address => Address,
            TypeTag::Signer => Signer,
            TypeTag::Vector(ty) => Vector(Box::new(Type::from(*ty))),
            TypeTag::Struct(s) => Struct {
                address: s.address,
                module: s.module,
                name: s.name,
                type_arguments: s.type_params.into_iter().map(|ty| ty.into()).collect(),
            },
        }
    }
}
```